Sources/Auth/Internal/SessionManager.swift

```diff
@@ -39,56 +43,67 @@ private actor LiveSessionManager {
 
   func session() async throws -> Session {
     try await trace(using: logger) {
-      guard let currentSession = try sessionStorage.get() else {
+      guard let currentSession = sessionStorage.get() else {
+        logger?.debug("session missing")
         throw AuthError.sessionMissing
       }
 
       if !currentSession.isExpired {
-        await scheduleNextTokenRefresh(currentSession)
-
         return currentSession
       }
 
+      logger?.debug("session expired")
       return try await refreshSession(currentSession.refreshToken)
     }
   }
 
   func refreshSession(_ refreshToken: String) async throws -> Session {
     try await SupabaseLoggerTaskLocal.$additionalContext.withValue(
-      merging: ["refreshID": .string(UUID().uuidString)]
+      merging: [
+        "refresh_id": .string(UUID().uuidString),
+        "refresh_token": .string(refreshToken),
+      ]
     ) {
       try await trace(using: logger) {
         if let inFlightRefreshTask {
-          logger?.debug("refresh already in flight")
+          logger?.debug("Refresh already in flight")
           return try await inFlightRefreshTask.value
         }
 
         inFlightRefreshTask = Task {
-          logger?.debug("refresh task started")
+          logger?.debug("Refresh task started")
 
           defer {
             inFlightRefreshTask = nil
-            logger?.debug("refresh task ended")
+            logger?.debug("Refresh task ended")
           }
 
-          let session = try await api.execute(
-            HTTPRequest(
-              url: configuration.url.appendingPathComponent("token"),
-              method: .post,
-              query: [
-                URLQueryItem(name: "grant_type", value: "refresh_token"),
-              ],
-              body: configuration.encoder.encode(UserCredentials(refreshToken: refreshToken))
+          do {
+            let session = try await api.execute(
+              HTTPRequest(
+                url: configuration.url.appendingPathComponent("token"),
+                method: .post,
+                query: [
+                  URLQueryItem(name: "grant_type", value: "refresh_token"),
+                ],
+                body: configuration.encoder.encode(
+                  UserCredentials(refreshToken: refreshToken)
+                )
+              )
             )
-          )
-          .decoded(as: Session.self, decoder: configuration.decoder)
+            .decoded(as: Session.self, decoder: configuration.decoder)
 
-          update(session)
-          eventEmitter.emit(.tokenRefreshed, session: session)
+            update(session)
+            eventEmitter.emit(.tokenRefreshed, session: session)
 
-          await scheduleNextTokenRefresh(session)
+            return session
+          } catch {
+            logger?.debug("Failed to refresh token: \(error)")
+            remove()
```

Comment:
Do we want to remove the token even if it's a network failure? Should we only remove it if it's a certain error code(s)?

Reply:
auth-js removes it if it isn't a retryable error https://github.com/supabase/auth-js/blob/master/src/GoTrueClient.ts#L1974-L1977

since the HTTP client retries the request a few times before reaching this point, the behavior is roughly the same as auth-js

How does flutter works?

Reply:
No, network failure means there's no internet. Users shouldn't be logged out ever from this.

Reply:
Ideally supplement the guides (and we should do this) to have the device detect there's likely no internet (Airplane mode, no Wifi, no Data / Cell service) and stop auto refreshes while this is true, as this would likely lead in a network error.

In the web it's a bit hard to detect this but I'll add it to auth-js.

Reply:
@hf in which cases should user be logged out? if any?

Reply:
I changed it so session is deleted only if error is non-retryable, just like JS and Flutter.